src/metabase/lib/stage.cljc

```diff
@@ -281,38 +298,60 @@
             (lib.metadata.calculation/implicitly-joinable-columns query stage-number existing-columns unique-name-fn)))
          vec)))
 
+(defn- column-signature [column-metadata]
+  (dissoc column-metadata
+          :source-alias :ident :lib/source :lib/source-uuid
+          :lib/desired-column-alias :lib/hack-original-name))
+
 ;;; Return results metadata about the expected columns in an MBQL query stage. If the query has
 ;;; aggregations/breakouts, then return those and the fields columns. Otherwise if there are fields columns return
 ;;; those and the joined columns. Otherwise return the defaults based on the source Table or previous stage + joins.
 (defmethod lib.metadata.calculation/returned-columns-method ::stage
-  [query stage-number _stage {:keys [unique-name-fn], :as options}]
+  [query stage-number _stage {:keys [include-remaps? unique-name-fn], :as options}]
+  #_(clojure.pprint/pprint ['RCM/stage _stage])
   (or
-   (existing-stage-metadata query stage-number)
-   (let [query        (ensure-previous-stages-have-metadata query stage-number)
-         summary-cols (summary-columns query stage-number unique-name-fn)
-         field-cols   (fields-columns query stage-number unique-name-fn)]
+   (existing-stage-metadata query stage-number unique-name-fn)
+   (let [query        (ensure-previous-stages-have-metadata query stage-number options)
+         summary-cols (summary-columns query stage-number options)
+         field-cols   (fields-columns query stage-number options)]
      ;; ... then calculate metadata for this stage
      (cond
        summary-cols
        (into summary-cols field-cols)
 
        field-cols
-       (do (doall field-cols)           ; force generation of unique names before join columns
-           (into []
-                 (m/distinct-by #(dissoc % :source-alias :ident :lib/source :lib/source-uuid :lib/desired-column-alias))
-                 (concat field-cols
-                         (lib.join/all-joins-expected-columns query stage-number options))))
+       (let [_          (doall field-cols)           ; force generation of unique names before join columns
+             join-cols  (lib.join/all-joins-expected-columns query stage-number options)
+             ;; The field-cols may contain would-be joined cols already! We de-duplicate them, but take the :ident
+             ;; from the first of the duplicates.
```

Comment:
Is "first" (i.e., the order) important? As I read the code, we take the ident of the _last_ of the duplicates.